File: 1.x/localsim/utils/signal.py
import collections
import weakref


class Signal(object):
    """Collects a dictionary of events connected to callbacks, and fires callbacks when prompted."""

    events = None
    __callbacks = None

    def connect(self, event, callback):
        if self.__callbacks is None:
            self.__callbacks = collections.defaultdict(weakref.WeakKeyDictionary)
        if event in self.events:
            if hasattr(callback, 'im_self'):
                weak_key = callback.im_self
                method_name = callback.__name__

                if weak_key not in self.__callbacks[event]:
                    self.__callbacks[event][weak_key] = [method_name]
                elif method_name not in self.__callbacks[event][weak_key]:
                    self.__callbacks[event][weak_key].append(method_name)

            else:
                raise TypeError('Callback not a method of a class')
        else:
            raise ValueError('Unknown event to this signal')

    def connect_to_all(self, callback):
        for event in list(self.events):
            self.connect(event, callback)

    def disconnect(self, event, callback):
        if event in self.events:
            if self.__callbacks:
                if hasattr(callback, 'im_self'):
                    weak_key = callback.im_self
                    method_name = callback.__name__

                    if weak_key in self.__callbacks[event] and method_name in self.__callbacks[event][weak_key]:
                        if len(self.__callbacks[event][weak_key]) == 1:
                            del self.__callbacks[event][weak_key]
                        else:
                            self.__callbacks[event][weak_key].remove(method_name)
                else:
                    raise TypeError('Callback not a method of a class')
        else:
            raise ValueError('Unknown event to this signal')

    def disconnect_to_all(self, callback):
        for event in list(self.events):
            try:
                self.disconnect(event, callback)
            except ValueError:
                pass

    def fire(self, event_, **kwargs):
        if event_ in list(self.events):
            if self.__callbacks:
                for weak_key, method_list in self.__callbacks[event_].items():
                    for m in list(method_list):
                        getattr(weak_key, m)(event_, self, **kwargs)

        else:
            raise ValueError('Unknown event to this signal')


class Clock(Signal):
    """Iterator based on the Signal object; acts as a clock."""

    events = ['fine', 'stop', 'coarse']

    def __init__(self, dt_fine, end=None, dt_coarse=None):
        self.start = 0
        self.end = end

        self.dt_fine = dt_fine
        self.dt_coarse = dt_coarse
        self.now = self.start

    def __iter__(self):
        return self

    def __next__(self):

        if not self.end or self.now <= self.end:
            self.now += self.dt_fine

            if self.dt_coarse and self.now % self.dt_coarse == 0:
                self.fire('coarse', time=self.now)

            self.fire('fine', time=self.now)

            return self.now
        else:
            self.fire('stop')
            raise StopIteration

    def __del__(self):
        self.fire('stop')

    def next(self):
        return self.__next__()

    def __repr__(self):
        return '<{klass}>: Time now is {now}. '.format(klass=self.__class__.__name__, now=self.now)

File: 1.x/localsim-web/utils/signal.js
/**
 * COMMENT
 * Base class for any object that takes in and fires callbacks
 */

class Signal{

    constructor(){
        this.events = null;
        this.__callbacks = null;
        this.__obj = null;

    }

    /**
     * Connects the callback to a Signal subclass object and a user-defined event
     * The callbacks are fired upon specifying an event in fire()
     * 
     * @param {Object} obj Object (whose class extends Signal) where the event comes from (?)
     * @param {String} event Key for certain user-defined actions
     * @param {Func} callback Values / Functions to be called when a specific event is specified
     */
    connect(obj, event, callback){
        this.__obj = obj;

        if(this.__callbacks == null){
            this.__callbacks = new Map();
        }


        if (this.events.includes(event)){
            if(callback instanceof Function){
                let weak_key = this.__obj;
                let method_name = callback.name;

                /**
                 * Assign to key 'event' a Map of (what will be) weak keys (the obj) to callback names
                 */
                if(!this.__callbacks.has(event)){
                    this.__callbacks.set(event, new Map()); // WeakMaps are not iterable
                }

                if(!(this.__callbacks.get(event).has(weak_key))){
                    this.__callbacks.get(event).set(weak_key, [method_name]);
                }

                else if(!(this.__callbacks.get(event).get(weak_key).includes(method_name))){
                    this.__callbacks.get(event).get(weak_key).push(method_name);
                }
            }

            else{
                throw 'Callback not a method of a class';
            }
        }

        else{
            throw 'Unknown event to this signal';
        }
    }

    connect_to_all(obj, callback){
        for(let ctr=0; ctr<this.events.length; ctr++){
            this.connect(obj, this.events[ctr], callback);
        }
    }

    disconnect(obj, event, callback){
        this.__obj = obj;
        if(this.events.includes(event)){
            if(this.__callbacks){
                if(callback instanceof Function){
                    let weak_key = this.__obj;
                    let method_name = callback.name;

                    if(this.__callbacks.get(event).has(weak_key) && this.__callbacks.get(event).get(weak_key).includes(method_name)){
                        if(this.__callbacks.get(event).get(weak_key).length == 1){
                            this.__callbacks.get(event).delete(weak_key);
                        }
                        else{
                            let index = this.__callbacks.get(event).get(weak_key).indexOf(method_name);
                            // Does this disconnect even work? index is never used
                        }
                    }
                }
            }
        }
    }

    disconnect_to_all(obj, callback){
        for(let ctr=0; ctr<this.events.length; ctr++){
            this.disconnect(obj, this.events[ctr], callback);
        }
    }

    /**
     * Fires all callbacks for an event_
     * 
     * @param {String} event_ The event triggered; whose callbacks will be fired
     * @param {Object} kwargs Keyword arguments
     */
    fire(event_, kwargs=null){
        let source = this;
        if(this.events.includes(event_)){
            if(this.__callbacks){
                this.__callbacks.get(event_).forEach(function(method_list, weak_key){
                    method_list.forEach(function(m){
                        weak_key[m](event_, source, kwargs);
                    });
                });
            }
        }
    }
}


class Clock extends Signal{
    constructor(dt_fine, end=null, dt_coarse=null){
        super();
        this.events = ['fine', 'stop', 'coarse'];

        this.start = 0;
        this.end = end;

        this.dt_fine = dt_fine;
        this.dt_coarse = dt_coarse;
        this.now = this.start;
    }

    next(){
        if(!this.end || this.now <= this.end){
            this.now += this.dt_fine;

            if(this.dt_coarse && (this.now % this.dt_coarse == 0)){
                this.fire('coarse', new Map().set('time', this.now));
            }

            this.fire('fine', new Map().set('time', this.now));

            return this.now;
        }
        else{
            this.fire('stop');
            throw 'StopIteration';
        }
    }

    stop(){
        this.now  = this.start;
        this.fire('stop');
    }

    print_curr_time(){
        console.log('Time is now ', this.now);
    }

    step(){
        let time = null;
        try{
            time = this.next()
        }
        catch(error){
            return null;
        }
        return time;
    }
}
